File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/headers/HeadersSpec.java
package io.micronaut.rabbitmq.docs.headers;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "HeadersSpec")
class HeadersSpec {

    @Test
    void testPublishingAndReceivingHeaders(ProductClient productClient, ProductListener productListener) {
// tag::producer[]
        productClient.send("body".getBytes());
        productClient.send("medium", 20L, "body2".getBytes());
        productClient.send(null, 30L, "body3".getBytes());

        Map<String, Object> headers = new HashMap<>(3);
        headers.put("productSize", "large");
        headers.put("x-product-count", 40L);
        productClient.send(headers, "body4".getBytes());
// end::producer[]

        await().atMost(60, SECONDS).until(() ->
                productListener.messageProperties.size() == 4 &&
                productListener.messageProperties.contains("true|10|small") &&
                productListener.messageProperties.contains("true|20|medium") &&
                productListener.messageProperties.contains("true|30|null") &&
                productListener.messageProperties.contains("true|40|large")
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/headers/HeadersSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="HeadersSpec")
class HeadersSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_publishing_and_receiving_headers(self) -> None:
# tag::producer[]
        self.product_client.send(b"body")
        self.product_client.send_with_headers("medium", 20, b"body2")
        self.product_client.send_with_headers(None, 30, b"body3")

        headers = {"productSize": "large", "x-product-count": 40}
        self.product_client.send_with_header_map(headers, b"body4")
# end::producer[]

        message_properties = self.product_listener.message_properties
        Await.until(lambda: len(message_properties) == 4
                    and "true|10|small" in message_properties
                    and "true|20|medium" in message_properties
                    and "true|30|None" in message_properties
                    and "true|40|large" in message_properties)
